### PR TITLE
python3Packages.cyipopt: init at 1.5.0

### PR DESCRIPTION
**Pull request by nim65s**
This add cython bindings to ipopt: https://github.com/mechmotum/cyipopt

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/development/python-modules/cyipopt/default.nix [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+  nativeBuildInputs = [ pkg-config ];
+
+  buildInputs = [ ipopt ];
+
+  build-system = [
+    cython
+    numpy
+    setuptools
+  ];
+
```
Comment (natsukium) [inline pkgs/development/python-modules/cyipopt/default.nix:33]: Could you add `numpy` to `dependencies`?
https://github.com/mechmotum/cyipopt/blob/9fe5b202d51294bfcc7b0d20d560eea49b2c344b/setup.py#L56-L58

File: pkgs/development/python-modules/cyipopt/default.nix [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  pythonImportsCheck = [ "cyipopt" ];
+
```
Comment (natsukium) [inline pkgs/development/python-modules/cyipopt/default.nix:35]: Could you enable tests with `pytestCheckHook`?

**Comment by nim65s** [comment]
Thanks for the review, @natsukium. I fixed both points, and correctly get :

`python3.12-cyipopt> ================== 30 passed, 63 skipped, 1 warning in 0.42s ===================`

File: pkgs/development/python-modules/cyipopt/default.nix [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+  nativeBuildInputs = [
+    pkg-config
+    pytestCheckHook
+  ];
+
+  buildInputs = [ ipopt ];
+
+  build-system = [
+    cython
+    numpy
```
Comment (FliegendeWurst) [inline pkgs/development/python-modules/cyipopt/default.nix:34]:
```suggestion
```

**Comment by FliegendeWurst** [comment]
@ofborg build python312Packages.cyipopt

File: pkgs/development/python-modules/cyipopt/default.nix (hunk 3 above)
Comment (natsukium) [reply pkgs/development/python-modules/cyipopt/default.nix]: Build-time dependencies and runtime dependencies are different, so numpy must be added to each input.
https://github.com/mechmotum/cyipopt/blob/9fe5b202d51294bfcc7b0d20d560eea49b2c344b/pyproject.toml#L2

File: pkgs/development/python-modules/cyipopt/default.nix [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+buildPythonPackage rec {
+  pname = "cyipopt";
+  version = "1.5.0";
+  pyproject = true;
+
+  src = fetchFromGitHub {
+    owner = "mechmotum";
+    repo = "cyipopt";
+    rev = "v${version}";
```
Comment (GaetanLepage) [inline pkgs/development/python-modules/cyipopt/default.nix:21]:
```suggestion
    rev = "refs/tags/v${version}";
```
To prevent branch/tag mismatch.

**Review by GaetanLepage — CHANGES_REQUESTED** [review]

**Comment by nim65s** [comment]
@GaetanLepage: thanks for the review, but I don't understand why I can't answer you directly in your comment

> Are there any tests in the end ?

Yes: `python3.12-cyipopt> ================== 30 passed, 63 skipped, 1 warning in 0.31s ===================`

File: pkgs/development/python-modules/cyipopt/default.nix (hunk 2 above)
Comment (nim65s) [reply pkgs/development/python-modules/cyipopt/default.nix:40]: Ok, that is here: yes: `python3.12-cyipopt> ================== 30 passed, 63 skipped, 1 warning in 0.31s ===================`

**Comment by FliegendeWurst** [comment]
> but I don't understand why I can't answer you directly in your comment

That is because Github duplicates review comments on existing threads. You can see their comment twice (once here, once at the original thread where you can answer)

File: pkgs/development/python-modules/cyipopt/default.nix (hunk 2 above)
Comment (GaetanLepage) [reply pkgs/development/python-modules/cyipopt/default.nix:40]: Oh right, they run automatically.
I think that it is better to explicitly add
```nix
nativeCheckInputs = [
  pytestCheckHook
];
```

Comment (nim65s) [reply pkgs/development/python-modules/cyipopt/default.nix:39]: Sure, it makes more sense, done, thanks.
And I still have this `python3.12-cyipopt> ================== 30 passed, 63 skipped, 1 warning in 0.31s ===================`

**Comment by GaetanLepage** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 349036`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 4 packages built:</summary>
  <ul>
    <li>python311Packages.cyipopt</li>
    <li>python311Packages.cyipopt.dist</li>
    <li>python312Packages.cyipopt</li>
    <li>python312Packages.cyipopt.dist</li>
  </ul>
</details>

---
### `aarch64-linux`
<details>
  <summary>:white_check_mark: 4 packages built:</summary>
  <ul>
    <li>python311Packages.cyipopt</li>
    <li>python311Packages.cyipopt.dist</li>
    <li>python312Packages.cyipopt</li>
    <li>python312Packages.cyipopt.dist</li>
  </ul>
</details>

---
### `x86_64-darwin`
<details>
  <summary>:white_check_mark: 4 packages built:</summary>
  <ul>
    <li>python311Packages.cyipopt</li>
    <li>python311Packages.cyipopt.dist</li>
    <li>python312Packages.cyipopt</li>
    <li>python312Packages.cyipopt.dist</li>
  </ul>
</details>

---
### `aarch64-darwin`
<details>
  <summary>:white_check_mark: 4 packages built:</summary>
  <ul>
    <li>python311Packages.cyipopt</li>
    <li>python311Packages.cyipopt.dist</li>
    <li>python312Packages.cyipopt</li>
    <li>python312Packages.cyipopt.dist</li>
  </ul>
</details>

**Review by GaetanLepage — APPROVED** [review]
LGTM

File: pkgs/development/python-modules/cyipopt/default.nix (hunk 3 above)
Comment (natsukium) [reply pkgs/development/python-modules/cyipopt/default.nix]: I'm not sure why this conversation was marked as resolved, but please add `numpy` to `build-system`.

**Pull request merged**